File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
n = int(input())
arr = list(map(int, input().split()))

LIS = [arr[0]]

def binarySearch(e):
    l, r = 0, len(LIS)-1
    while l <= r:
        mid = (l+r)//2
        if LIS[mid] == e:
            return mid
        elif LIS[mid] < e:
            l = mid + 1
        else:
            r = mid - 1
    return l

for i in arr:
    if LIS[-1] < i:
        LIS.append(i)
    else:
        idx = binarySearch(i)
        LIS[idx] = i

print(len(LIS))

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

var lis []int
var arr []int
var n int

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	temp, _ := reader.ReadString('\n')
	n, _ = strconv.Atoi(strings.TrimSpace(temp))

	temp, _ = reader.ReadString('\n')
	inputs := strings.Fields(temp)
	arr = make([]int, n)
	for index, value := range inputs {
		arr[index], _ = strconv.Atoi(value)
	}

	lis = []int{arr[0]}
	for i := 0; i < n; i++ {
		if arr[i] > lis[len(lis)-1] {
			lis = append(lis, arr[i])
		} else {
			idx := binarySearch(arr[i])
			lis[idx] = arr[i]
		}
	}

	writer.WriteString(strconv.Itoa(len(lis)))
}

func binarySearch(target int) int {
	left, right := 0, len(lis)-1
	for left <= right {
		mid := (left + right) / 2
		if lis[mid] == target {
			return mid
		} else if lis[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return left
}